cirq/neutral_atoms/neutral_atom_devices.py

```diff
@@ -0,0 +1,426 @@
+
+    def validate_operation(self, operation: ops.Operation):
+        """
+        Raises an error if the given operation is invalid on this device.
+
+        Args:
+            operation: the operation to validate
+
+        Raises:
+            ValueError: If the operation is not valid
+        """
+        if not isinstance(operation, (ops.GateOperation,
+                                      ops.ParallelGateOperation)):
+            raise ValueError('Unsupported operation: {!r}'.format(operation))
+
+        # The gate must be valid
+        self.validate_gate(operation.gate)
+
+        # All qubits the operation acts on must be on the device
+        for q in operation.qubits:
+            if q not in self.qubits:
+                raise ValueError('Qubit not on device: {!r}'.format(q))
+
+        if isinstance(operation.gate, (ops.MeasurementGate, ops.IdentityGate)):
+            return
+
+        # Verify that a controlled gate operation is valid
+        if isinstance(operation, ops.GateOperation):
+            if len(operation.qubits) > self._max_parallel_c:
+                raise ValueError("Too many qubits acted on in parallel by a "
+                                 "controlled gate operation")
+            if len(operation.qubits) > 1:
+                for p in operation.qubits:
+                    for q in operation.qubits:
+                        if self.distance(p, q) > self._control_radius:
+                            raise ValueError("Qubits {!r}, {!r} are too "
+                                             "far away".format(p, q))
+            return
+
+        # Verify that a valid number of Z gates are applied in parallel
+        if isinstance(operation.gate, ops.ZPowGate):
+            if len(operation.qubits) > self._max_parallel_z:
+                raise ValueError("Too many Z gates in parallel")
+
+        # Verify that a valid number of XY gates are applied in parallel
+        if isinstance(operation.gate, (ops.XPowGate,
+                                       ops.YPowGate,
+                                       ops.PhasedXPowGate)):
+            if (len(operation.qubits) > self._max_parallel_xy and
+                    len(operation.qubits) != len(self.qubits)):
+                raise ValueError("Bad number of XY gates in parallel")
+
+    def validate_moment(self, moment: ops.Moment):
+        """
+        Raises an error if the given moment is invalid on this device
+
+        Args:
+            moment: The moment to validate
+
+        Raises:
+            ValueError: If the given moment is invalid
+        """
+        super().validate_moment(moment)
+
+        operations = moment.operations
+        num_parallel_z = 0
+        num_parallel_xy = 0
+        has_measurement = False
+        controlled_qubits_lists = []  # type: List[Tuple[raw_types.Qid, ...]]
+
+        # Count the number of each type of gate occuring in this moment.
```

Comment:
This method is getting closer to reasonable. What do you think of this change?

```python
    categorized_ops = collections.defaultdict(list)
    for op in operations:
        assert isinstance(op, (ops.GateOperation, ops.ParallelGateOperation))
        categorized_ops[GATE_CATEGORIES[type(op.gate)]].append(op)

    for k in ['Z', 'XY', 'controlled']:
        if len(set(op.gate for op in categorized_ops[k])) > 1:
            raise ValueError("Non-identical simultaneous {} gates".format(k))

    num_parallel_xy = len(categorized_ops['XY'])
    num_parallel_z = len(categorized_ops['Z'])
    has_measurement = len(categorized_ops['measure']) > 0
    controlled_qubits_lists = [op.qubits
                               for op in categorized_ops['controlled']]
```

where

```python
    GATE_CATEGORIES = {
        ops.ZPowGate: 'Z',

        ops.XPowGate: 'XY',
        ops.YPowGate: 'XY',
        ops.PhasedXPowGate: 'XY',

        ops.CNotPowGate: 'controlled',
        ops.CZPowGate: 'controlled',
        ops.CCXPowGate: 'controlled',
        ops.CCZPowGate: 'controlled',

        ops.MeasurementGate: 'measure',
    }
```

Reply:
I like this change. The num_parallel_xy and num_parallel_z variables need to be calculated differently to account for the existence of ParallelGateOperations to something like `num_parallel_xy = sum([len(op.qubits) for op in categorized_ops['XY']])`. I'll implement this in my next commit.

Reply:
I also needed to add a few more types into the dictionary to catch the Pauli types, but this worked and is now implemented.